Triangulation weights callback

gassmoeller opened the pull request:
This is an enhancement that evolved out of a discussion on the ASPECT github page: https://github.com/geodynamics/aspect/pull/617
We would like to be able to assign repartition weights to cells during refinement_and_coarsening, because we want to avoid to transfer our solution and attached data twice (once for the refinement, and once for the repartitioning). It is not possible to simply hand the weights over to ´execute_coarsening_and_refinement`, because we do not know yet, how the refinement will change the mesh. It would be possible however, with a signal that calls a callback_function for every cell, which hands over a cell reference and what will happen to this cell, and returns a weight for this particular cell.

This is work in progress in that not all comments are up to date, and the partitioning in some of my tests look strange, although the general behaviour looks correct. I have 2 questions though, and would like some feedback:
- currently, we only define signals for `Triangulation`, but this new signal is specific for parallel::Triangulation. For now I copied the Signals structure into `parallel::distributed::Triangulation`, but maybe it would be better to start something like a parallel::Triangulation::ParallelSignals?
- There is already a class `PartitionWeights`, which can store and provide the cell weights for p4est. However, this class is designed for taking a list of cells ordered in deal.II ordering and sorting it into p4est order. Because the new approach can not iterate in deal.II order (some cells are not there yet), I already have a list of weights sorted in p4est order. For now I copied PartitionWeights and stripped away the sorting, but of course I could also change PartitionWeights to take an additional parameter that indicates, whether it should do the sorting (maybe with the default set to do the sorting, to keep compatibility?). Would that be preferrable to remove code duplication?

tjhei commented:
> For now I copied the Signals structure

I am not even sure your approach will work. Your ```p::d::Triangulation``` now has two members called ```signal``` and therefore two signals called ```pre_refinement```.

bangerth commented:
The correct way would then be to have the base class not store a `Signals` object, but a `std_cxx11::unique_ptr<Signals>` pointer. The derived class can then have a derived `Signals` class and reset the pointer stored in the base class to a pointer-to-derived.

bangerth commented on include/deal.II/distributed/tria.h:
```diff
@@ -818,6 +818,55 @@ namespace parallel
+        boost::signals2::signal<void ()> clear;
+        boost::signals2::signal<void ()> any_change;
+        boost::signals2::signal<unsigned int (const cell_iterator &,
+                                              const CellStatus),Triangulation<dim,spacedim>::sum<unsigned int> > cell_weight;
```
Put the accumulator template argument onto its own line to make it easier to read.

bangerth commented on source/distributed/tria.cc:
```diff
@@ -1540,6 +1540,92 @@ namespace
+      }
+    else
+      {
+        //it's children got coarsened into this cell
```
it's -> its

gassmoeller commented:
@tjhei you are right, I have not thought about that (just tested the new signal so far).
 
But Wolfgang would your suggestion not break backwards compatibility? Every current user code `triangulation.signals.pre_refinement.connect()` would need to be replaced by `triangulation.signals->pre_refinement.connet()`. I certainly agree that it would be the more extensible approach, I am just not sure if we want to do this, or if there is better way?

gassmoeller commented:
This should now work as intended, although it is not the prettiest solution. 
parallel::distributed::Triangulation now has an additional parallel_signals object that contains the cell_weight signal. @bangerth: do you still prefer the unique_ptr? It should be simple to change that, but I am afraid of the consequences ;-)

bangerth commented:
On 10/05/2015 09:31 AM, Rene Gassmöller wrote:
> But Wolfgang would your suggestion not break backwards compatibility?
> Every current user code |triangulation.signals.pre_refinement.connect()|
> would need to be replaced by
> |triangulation.signals->pre_refinement.connet()|. I certainly agree that
> it would be the more extensible approach, I am just not sure if we want
> to do this, or if there is better way?

Right, crap. I knew that using a public member variable was a risk and 
did it anyway :-(

bangerth commented on include/deal.II/distributed/tria.h:
```diff
@@ -818,6 +818,50 @@ namespace parallel
+      {
+        boost::signals2::signal<unsigned int (const cell_iterator &,
+                                              const CellStatus),
+                                                    Triangulation<dim,spacedim>::sum<unsigned int> > cell_weight;
```
Can you document this signal somewhere?

bangerth commented on include/deal.II/distributed/tria.h:
```diff
@@ -953,6 +997,13 @@ namespace parallel
+       * weights before repartitioning occurs. Called from
+       * execute_coarsening_and_refinement().
+       */
+      void get_cell_weights(std::vector<unsigned int> &weights);
```
Can you document the intent of the argument (e.g., its size, what it represents, etc)?

bangerth commented on source/distributed/tria.cc:
```diff
@@ -3582,21 +3686,35 @@ namespace parallel
+      // before repartitioning get cell weights for a weighted repartitioning.
+      // TODO: move this call into the if below, only call when slots are connected
+      std::vector<unsigned int> cell_weights;
+      get_cell_weights(cell_weights);
```
In recent times, we've gone more towards making functions like this actually return the object, rather than taking the return object as last argument. Can you change this so that `get_cell_weights` returns the vector? This also allows you to assign it to a `const` vector.

bangerth commented on source/distributed/tria.cc:
```diff
@@ -4532,6 +4650,50 @@ namespace parallel
+    {
+      Assert(weights.size() == 0,
+             ExcMessage("You are trying to calculate weights, but the given vector is not empty. Refusing to overwrite content."));
+      weights.reserve(this->n_active_cells());
```
ah, nice, I had already wondered whether you'd do this `reserve` :-)

I think you don't actually need `n_active_cells()` slots, though. Aren't you only going to push `n_locally_owned_active_cells()` into the vector?

bangerth commented:
Looks pretty good, only a few small issues.

Can you write some documentation that documents this approach? I think if you look for the word "partition" in `include/deal.II/distributed/tria.h`, you will likely find all of the instances where there is already documentation that might have to be updated.

tjhei commented:
> Right, crap. I knew that using a public member variable was a risk and did it anyway :-(

Should we fix this by deprecating the current structure, introducing a new function ```Signals & get_signals()```, and overriding this inside d::p::Tria by returning a structure ParallelSignals derived from it. That would mean we have two signals we have to call until we remove the deprecated struct.

bangerth commented:
Yes, I thought about something like this. It's a bit of a pain with the deprecation, but it's probably the right thing to do.

We could avoid the issue with the two signals if we just moved the signals for the parallel class into the `Signals` structure of the base class until we break backward compatibility and remove the public variable (at which point we can use the `unique_ptr` thingy).

tjhei commented:
I wouldn't mind having a ```weight``` signal in the base class that does nothing if not used with a p::d::Tria. Easy enough to document.

bangerth commented:
True. It may even be of use to `shared::Triangulation`.

gassmoeller commented:
I moved the signal (and its dependencies) into Triangulation and added documentation for the new signal. Are you ok with this version? Should we add the deprecation warning and get_signals() in a separate PR?

tjhei commented:
Don't worry about the ```get_signals()``` function and deprecation for this PR. My vote is for keeping everything as is anyways.

Can you please check/do:
- update the documentation of ```execute_coarsening_and_refinement``` (you can weigh the cells, refer to the signal)
- deprecate the old way of setting weights in ```repartition()``` by removing the default argument and adding a new overload ```repartition(std::vector)``` without default value that is deprecated. Okay, @bangerth ?
- update the documentation of ```repartition()```
- use the signals in the new ```repartition()``` function without arguments
- do we need to call ```repartition()``` instead of ```partition()``` in ```load()``` now so that signals work? (https://github.com/gassmoeller/dealii/blob/triangulation_weights_callback/source/distributed/tria.cc#L2867)

bangerth commented:
On 10/07/2015 12:24 PM, Timo Heister wrote:
> # deprecate the old way of setting weights in |repartition()| by removing
> the default argument and adding a new overload
> |repartition(std::vector)| without default value that is deprecated.
> Okay, @bangerth <https://github.com/bangerth> ?
> #

I don't feel strongly about the old function. I think it might be 
simpler at times to use the old interface in some cases because it is 
simpler to fill a vector than to write a callback/slot. But I figure 
that the number of people who use it is small and they know that they 
need to do advanced things.

gassmoeller commented:
Would it be confusing to just keep both versions of repartition()? I would not even need to create a new function just check:
- if a vector of weights is provided: use it
- if not, but at least one slot is connected to the signal: use it
- if none of the above balance number of cells as normal

tjhei commented:
The reason to split it into two functions is so that you can deprecate the one with the ```std::vector``` that it will generate a warning for the user. If we decide to deprecate it that is. You also need to handle the case where both things are given, btw.

bangerth commented:
I like the simplicity of the current system. Couldn't we just have a system whereby `repartition()` simply adds what it gets from the signal to the values it already gets as argument? The existing argument's default value is a zero vector, so this would nicely generalize.

bangerth commented:
I think @gassmoeller 's proposal above actually already implements what I just said, except that I don't think that an `if` statement is needed to distinguish between the first two cases if we add things up. A signal with no attached slots and an integer combiner will simply return zero, which you can safely add to the weights from the vector.

gassmoeller commented:
I agree. My line of thought was mostly that both systems do not conflict, and the old system is easier to use if you only want to use repartition. I am fine with adding the possibility to sum vector value and signal return value, the only complication is that they are sorted differently. Only the constructor of PartitionWeights sorts the vector values in the same order as the signal values. But I could write an operator + for PartitionWeights (which simply loops over all now sorted cell_weights and sums them), and then use `PartitionWeights summed_weights = vector_weights + signal_weights`. @tjhei: Would you agree with this approach, or do you prefer to deprecate the old version to have only one way of providing weights?

tjhei commented:
I would prefer to deprecate the old method for the following reasons:
1. Having nearly the same functionality implemented in two different ways is confusing (we need to document when each method is in effect and what happens if you apply both).
2. The new one is more correct (because you can set weights for cells that will be refined/coarsened correctly).
3. The new method doesn't require this awkward ``no_automatic_repartitioning``` which messes up SolutionTransfer. 
4. The new method can also work for initial distribution, ```load()```, etc.

tjhei commented:
> I am fine with adding the possibility to sum vector value and signal return value, the only complication is that they are sorted differently

You could implement the old method by attaching to the signal and looking up the weight from the vector (something like ```return weight_vector[cell->index()]```).

bangerth commented:
>You could implement the old method by attaching to the signal and
>looking up the weight from the vector (something like ```return
>weight_vector[cell->index()]```).

Yes, that's actually quite elegant.

bangerth commented:
It's `cell-> active_cell_index ()` in the square brackets, actually.

tjhei commented:
Of course, that's what I meant. ;-)

bangerth commented:
I don't disagree with @tjhei 's general statement about deprecating the old behavior. But, I find that the effort to write a separate function as a slot for a signal is so much larger than the current interface that it might make sense considering keeping the old function.

@tjhei : Do you propose to break backward compatibility right away? If not, then we have to have both methods working at the same time anyway, and presumably also document how they interact. This might give us a bit of time to see how this new interface actually works before we have to make a decision about deprecating the old interface.

gassmoeller commented:
For now I separated the two methods as much as possible to make deprecating the old interface easy (but did not do it yet). To deprecate/remove the old functionality one would only need to change the following functions:
- repartition(cell_weights) 
- one of the constructors of PartitionWeights (the one with the many arguments)
- PartitionWeights::build_weights_list.

The current state is something like a 'soft' deprecation. There is no warning when using the old function, but it is not mentioned in the documentation (that only refers to the new functionality). In the current state there is no interaction between the two ways (you either use the signal, or the provided vector).

tjhei commented on source/distributed/tria.cc:
```diff
@@ -1540,6 +1540,86 @@ namespace
+      {
+        //this cell got refined, so get the weight for parent and divide it
+        unsigned int parent_weight(GeometryInfo<dim>::max_children_per_cell * 1000);
+        parent_weight += signals.cell_weight(dealii_cell,
```
You are not multiplying the return value with ```max_children_per_cell``` here, so this means the interpretation of the return value is "cost adjustment for one of the children of this cell that is being refined". Is that intuitive? I think the typical case is that something like:
```
if (cell_is_in_expensive_region)
  return 1000; // require 2x the work
else
  return 0;
```
which won't do what you expect if ```CELL_REFINE``` is set. Thoughts?

gassmoeller replied:
Actually for my use case I think this is exactly what I expect. I have a certain weight for the parent cell (let's say some markers to move around), and when I refine that cell every child only gets 1000 plus one quarter of the additional weight (in 2d). But I see that for other use cases (like a more expensive equation to be solved on this cell) the expected behaviour would be different. I do not feel strongly about what way is better, should we go with your way and simply document it so that users know?

tjhei replied:
I don't care too much either way, but it needs to be documented for sure!

I have been thinking about this quite a bit over the last couple of days and it is hard to wrap my head around it. I guess it boils down to the question if the extra work will be split when refining (your case where each child will have n/4 particles) or not (say you have an hp element that requires more expensive assembly). What is the more common case? I don't know.

bangerth replied:
I think the more common case is where the work on each of the children is about as much as on the parent, as opposed to 1/4 (or 1/8th) of it.

tjhei commented:
> @tjhei : Do you propose to break backward compatibility right away? If not, then we have to have both methods working at the same time anyway, and presumably also document how they interact. This might give us a bit of time to see how this new interface actually works before we have to make a decision about deprecating the old interface.

I am fine with that.

tjhei commented:
One more thought: the old weight functionality was introduced after 8.3 so we could also just outright remove it if that makes things easier. @bangerth ?

bangerth commented:
Correct, we never released that function.

@tjhei : I defer to you how you want to proceed with this. You seem to feel much more strongly about getting rid of the old approach.

gassmoeller commented:
I changed the handling of child cells as discussed, and moved some documentation around to be more in line with what #1743 will do. The only missing part is now the handling of the old functionality: Keep, deprecate or remove? For reference: The old functionality was introduced by Wolfgang in mid July. I do not know how many people use it since then, or for what purpose it was introduced, so I would also just follow your decision @tjhei ;-)

tjhei commented:
Looks good to me! Thanks for incorporating all the changes. I think the new formulation is more flexible and cleaner, so I would just remove the old version.

gassmoeller commented:
Ok this should finally address all remaining issues. The old functionality repartition(cell_weights) has been removed, and all cell_weights tests now use the new functionality. If you are fine with this version, feel free to merge.

tjhei commented:
Thanks, @gassmoeller. Looks good. Can we get a changelog entry?

Any final comments, @bangerth?

bangerth commented:
On 10/31/2015 02:19 PM, Timo Heister wrote:
> Any final comments, @bangerth <https://github.com/bangerth>?

Nope :-)

gassmoeller commented:
I added the entry and rebased to current master. If the build tester is happy this is ready for merge.

bangerth commented:
It looks like the merge of this PR broke test `mpi/p4est_save_04.mpirun=5.debug`, see http://cdash.kyomu.43-1.org/testDetails.php?test=25606822&build=14125

@gassmoeller -- if you run this locally for you, do you see the same?

gassmoeller commented:
I could not test, because I did not had petsc (currently compiling), but I am pretty certain what causes the problem. Previously we just loaded the triangulation after a save(), now we also repartition it again (only important in case the number of processors has changed, at least that is what the comment there says). This triggers an assertion that repartition() should only be called when the `no_automatic_repartitioning` flag is set, which is not done in this test. 

However, with the new functionality, I do not see a reason why we should not allow users to have an automatic repartitioning, and still call repartition() themselves when they like to (e.g. because something has unexpectedly changed). So in case you agree and the tests pass, I would just remove the assertion.
